File: sources/builders/naptan.ts
/**
 * Build NaPTAN stops database
 *
 * Usage: bun run sources/builders/naptan.ts
 */

/* biome-ignore-all lint/suspicious/noConsole: Build script uses console for progress */

import { Database } from 'bun:sqlite';
import { createReadStream, mkdirSync } from 'node:fs';
import { join } from 'node:path';
import { createInterface } from 'node:readline';
import { createBatchInserter, downloadFile, findColumnIndex, parseCsvLine, progress, progressDone, withTempDir } from '../utils/index.ts';

/**
 * Source page: https://www.data.gov.uk/dataset/ff93ffc1-6656-47d8-9155-85ea0b8f2251/naptan
 * Direct download: https://naptan.api.dft.gov.uk/v1/access-nodes?dataFormat=csv
 */
const NAPTAN_CSV_URL = 'https://naptan.api.dft.gov.uk/v1/access-nodes?dataFormat=csv';
const DB_PATH = join(import.meta.dirname, '..', 'db', 'naptan.db');
mkdirSync(join(import.meta.dirname, '..', 'db'), { recursive: true });

type StopRow = [string, string | null, string | null, string | null, number | null, number | null];

interface ColumnIndices {
	atco: number;
	naptan: number;
	name: number;
	type: number;
	lat: number;
	lng: number;
}

function parseHeader(line: string): ColumnIndices {
	const header = parseCsvLine(line);
	const indices: ColumnIndices = {
		atco: findColumnIndex(header, ['atco', 'code']),
		naptan: findColumnIndex(header, ['naptan', 'code']),
		name: findColumnIndex(header, ['common', 'name']),
		type: findColumnIndex(header, ['stop type']),
		lat: findColumnIndex(header, ['latitude']),
		lng: findColumnIndex(header, ['longitude']),
	};
	if (indices.atco < 0 || indices.lat < 0 || indices.lng < 0) {
		throw new Error('Required NaPTAN columns not found');
	}
	return indices;
}

function optionalColumn(cols: string[], idx: number): string | null {
	if (idx < 0) return null;
	return cols[idx]?.trim() ?? null;
}

function parseCoordinate(raw: string | undefined): number | null {
	if (!raw) return null;
	const value = Number.parseFloat(raw);
	return Number.isNaN(value) ? null : value;
}

function parseRow(cols: string[], idx: ColumnIndices): StopRow | null {
	const atco = cols[idx.atco]?.trim();
	if (!atco) return null;
	return [atco, optionalColumn(cols, idx.naptan), optionalColumn(cols, idx.name), optionalColumn(cols, idx.type), parseCoordinate(cols[idx.lat]), parseCoordinate(cols[idx.lng])];
}

async function buildDatabase(csvPath: string): Promise<void> {
	console.log('Building NaPTAN database...');
	console.log(`Source: ${csvPath}`);
	console.log(`Output: ${DB_PATH}\n`);

	const db = new Database(DB_PATH, { create: true });
	db.exec(`
		DROP TABLE IF EXISTS stops;
		CREATE TABLE stops (
			atco_code TEXT PRIMARY KEY,
			naptan_code TEXT,
			common_name TEXT,
			stop_type TEXT,
			lat REAL,
			lng REAL
		);
		CREATE INDEX idx_stops_lat_lng ON stops(lat, lng);
	`);

	const inserter = createBatchInserter<StopRow>(db, 'INSERT INTO stops (atco_code, naptan_code, common_name, stop_type, lat, lng) VALUES (?, ?, ?, ?, ?, ?)', {
		batchSize: 10000,
		onProgress: (n) => progress(`Processed ${n.toLocaleString()} rows`),
	});

	const stream = createReadStream(csvPath, { encoding: 'utf-8' });
	const rl = createInterface({ input: stream, crlfDelay: Infinity });

	let idx: ColumnIndices | null = null;

	for await (const line of rl) {
		if (!idx) {
			idx = parseHeader(line);
			continue;
		}

		const row = parseRow(parseCsvLine(line), idx);
		if (row) inserter.add(row);
	}

	const totalRows = inserter.flush();
	progressDone();
	console.log(`Inserted ${totalRows.toLocaleString()} stops.`);
	db.close();
}

export async function build(): Promise<void> {
	await withTempDir('naptan', async (tempDir) => {
		const csvPath = join(tempDir, 'naptan.csv');
		console.log('Downloading NaPTAN stops...');
		console.log(`URL: ${NAPTAN_CSV_URL}\n`);
		await downloadFile(NAPTAN_CSV_URL, csvPath);
		console.log('Download complete.\n');
		await buildDatabase(csvPath);
	});
}

if (import.meta.main) {
	build().catch((err) => {
		console.error('NaPTAN build failed:', err);
		process.exit(1);
	});
}

File: crates/let-sdk/src/sources/naptan.rs
#![forbid(unsafe_code)]

use std::path::Path;

use csv::{ReaderBuilder, StringRecord};

use crate::errors::{ErrorCode, LetError, Result};

use super::common::{download_file, find_column_index, open_source_db, to_f64, with_temp_dir};

const NAPTAN_CSV_URL: &str = "https://naptan.api.dft.gov.uk/v1/access-nodes?dataFormat=csv";

pub fn build(db_path: &Path) -> Result<usize> {
    let temp = with_temp_dir()?;
    let csv_path = temp.path().join("naptan.csv");
    download_file(NAPTAN_CSV_URL, &csv_path)?;

    let mut reader = ReaderBuilder::new()
        .has_headers(true)
        .flexible(true)
        .from_path(&csv_path)
        .map_err(|error| {
            LetError::new(
                ErrorCode::Parse,
                format!("failed to open naptan CSV: {error}"),
                "verify naptan source file format",
            )
        })?;

    let headers = reader
        .headers()
        .map_err(|error| {
            LetError::new(
                ErrorCode::Parse,
                format!("failed to read naptan CSV headers: {error}"),
                "verify naptan source file format",
            )
        })?
        .iter()
        .map(|value| value.trim().to_lowercase())
        .collect::<Vec<_>>();

    let atco = find_column_index(&headers, &["atco", "code"]).ok_or_else(|| {
        LetError::new(
            ErrorCode::Parse,
            "missing ATCO code column in naptan CSV".to_owned(),
            "verify naptan source file format",
        )
    })?;

    let lat = find_column_index(&headers, &["latitude"]).ok_or_else(|| {
        LetError::new(
            ErrorCode::Parse,
            "missing latitude column in naptan CSV".to_owned(),
            "verify naptan source file format",
        )
    })?;

    let lng = find_column_index(&headers, &["longitude"]).ok_or_else(|| {
        LetError::new(
            ErrorCode::Parse,
            "missing longitude column in naptan CSV".to_owned(),
            "verify naptan source file format",
        )
    })?;

    let naptan = find_column_index(&headers, &["naptan", "code"]);
    let common_name = find_column_index(&headers, &["common", "name"]);
    let stop_type = find_column_index(&headers, &["stop type"]);

    let mut connection = open_source_db(db_path)?;
    connection.execute_batch(
        "
        DROP TABLE IF EXISTS stops;
        CREATE TABLE stops (
            atco_code TEXT PRIMARY KEY,
            naptan_code TEXT,
            common_name TEXT,
            stop_type TEXT,
            lat REAL,
            lng REAL
        );
        CREATE INDEX idx_stops_lat_lng ON stops(lat, lng);
        ",
    )?;

    let tx = connection.transaction()?;
    let mut statement = tx.prepare(
        "INSERT INTO stops (atco_code, naptan_code, common_name, stop_type, lat, lng) VALUES (?1, ?2, ?3, ?4, ?5, ?6)",
    )?;

    let mut inserted = 0usize;
    for row in reader.records() {
        let row = row.map_err(csv_err)?;
        let Some(atco_code) = cell(&row, atco) else {
            continue;
        };

        statement.execute(rusqlite::params![
            atco_code,
            optional_cell(&row, naptan),
            optional_cell(&row, common_name),
            optional_cell(&row, stop_type),
            to_f64(cell(&row, lat)),
            to_f64(cell(&row, lng)),
        ])?;

        inserted += 1;
    }

    drop(statement);
    tx.commit()?;
    connection.execute_batch("VACUUM; ANALYZE;")?;

    Ok(inserted)
}

fn cell(row: &StringRecord, idx: usize) -> Option<&str> {
    row.get(idx)
        .map(str::trim)
        .and_then(|value| (!value.is_empty()).then_some(value))
}

fn optional_cell(row: &StringRecord, idx: Option<usize>) -> Option<&str> {
    idx.and_then(|index| cell(row, index))
}

fn csv_err(error: csv::Error) -> LetError {
    LetError::new(
        ErrorCode::Parse,
        format!("failed to parse naptan CSV row: {error}"),
        "verify naptan source file format",
    )
}
